### ISSUE TITLE
Error launching demos (Continuing issue #20)

### ISSUE THREAD

**Issue by RicardoRagel** [ethz-asl/aerial_mapper#22]
Continuing issue #20, and copying the last comment:

Hi and thanks for the changes and the support @hitimo!

I have seen the changes in the last commit and in the installation instructions. There is only one problem: In the installation instructions, maybe the order must be: First `wstool update -j8` and later `sed -i ... opencv3_catkin/CMakeLists.txt`

The `0-synthetic-cadastre-dsm.launch` works perfectly now! The accumulated point cloud looks very nice :)

On the other hand, the `1-real-rafz-ortho-from-homography.launch` looks like it that does not work correctly. When I launch it, the rafz_gray dataset is load correctly and the node shows the sequence of images, but no homography result is shown and the program fails in:

```
I0122 10:23:54.483240  4905 aerial-mapper-io.cc:226] Number of images loaded: 249
*** Aborted at 1516613034 (unix time) try "date -d @1516613034" if you are using GNU date ***
PC: @     0x7f4af791e712 ortho::OrthoForwardHomography::prepareBlenderForNextImage()
*** SIGSEGV (@0x0) received by PID 4905 (TID 0x7f4af7cd6800) from PID 0; stack trace: ***
    @     0x7f4af54a7cb0 (unknown)
    @     0x7f4af791e712 ortho::OrthoForwardHomography::prepareBlenderForNextImage()
    @     0x7f4af791f75b ortho::OrthoForwardHomography::OrthoForwardHomography()
    @           0x402361 main
    @     0x7f4af5492f45 __libc_start_main
    @           0x4031f7 (unknown)
[demo_ortho_forward_homography-2] process has died
```
Maybe it has a similar opencv problem?

Thanks and great work!

**Comment by yjmenezes** [comment]
@RicardoRagel 

I am running ros/indigo under a QEMU guest Ubuntu 14.04x64  ram=30GB,  AMD 6 cores.

btw, the /tmp/simulation directory has two text files.
a camera_fixed_wing.yaml and estimated 249 UTM body poses: opt_poses.txt.
I plot opt_poses.txt UTM and it  looks ok. ( like an pose ordered arc ).

0-synthetic-cadastre-dsm.launch works for me too.
1-real-rafz-ortho-from-homography.launch   crash as you pointed out. same messages.
I run roslaunch -v  

You are right, the problem seems to be with OpenCV.
Have you tried other dataset ?


My messages:

started core service [/rosout]
process[demo_ortho_forward_homography-2]: started with pid [29148]
I0128 17:13:38.596346 29148 aerial-mapper-io.cc:255] Loading camera calibration from: /tmp/simulation/camera_fixed_wing.yaml
W0128 17:13:38.597270 29148 ncamera-yaml-serialization.cc:30] Unable to get the id for the ncamera. Generating new random id.
W0128 17:13:38.597502 29148 camera-yaml-serialization.cc:131] Unable to get the id for the camera. Generating new random id.
I0128 17:13:38.597612 29148 aerial-mapper-io.cc:40] Loading body poses via format: 0
I0128 17:13:38.597625 29148 aerial-mapper-io.cc:107] Loading body poses from: /tmp/simulation/opt_poses.txt
I0128 17:13:38.598642 29148 aerial-mapper-io.cc:120] T_G_Bs->size() = 249
I0128 17:13:38.598657 29148 aerial-mapper-io.cc:211] Loading images from directory+prefix: /tmp/simulation/opt_image_
OpenCV Error: Assertion failed (size.width>0 && size.height>0) in imshow, file /home/yjmenezes/catkin_ws_aerial_mapper/build/opencv3_catkin/opencv3_src/modules/highgui/src/window.cpp, line 312
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/yjmenezes/catkin_ws_aerial_mapper/build/opencv3_catkin/opencv3_src/modules/highgui/src/window.cpp:312: error: (-215) size.width>0 && size.height>0 in function imshow

yes, thx @hitimo for his great work !

**Comment by RicardoRagel** [comment]
Buenos días @yjmenezes

I think that trying other datasets will fail into the same OpenCV problem, so I have not tested any other dataset. I am waiting for @hitimo fix this problem (please ;)) to continue testing it. 

Thx!

**Comment by RicardoRagel** [comment]
Good morning @yjmenezes !

Did you execute (installation instructions):

FIRST: `wstool update -j8`
SECOND: `sed -i 's/WITH_QT=OFF/WITH_QT=ON/g' opencv3_catkin/CMakeLists.txt`

Reading your command line error, the program fails at (opencv::) imgshow(), but opencv is in the rosinstall dependencies, so I think it was already installed in your computer.

**Comment by yjmenezes** [comment]
@RicardoRagel 

Yes, i swapped sed line position.
I wanna repeat install over a fresh Ubuntu.
I saw some ldconfig messages about opencv version conflict.
I wanna check it.

**Comment by hitimo** [comment]
Hi @RicardoRagel 
- Thanks a lot for reporting the fix for the installation instructions, this is corrected now.
- I cannot reproduce your error: `roslaunch 1-real-rafz-ortho-from-homography.launch` shows the images and generates the orthomosaic which it stores to `/tmp/result.jpg` by default. So `xdg-open /tmp/result.jpg ` should show you the mosaic.
- Please make sure that the executable pulled in only one opencv version, i.e. ```ldd ~/catkin_ws/devel/lib/libaerial_mapper_demos.so | grep opencv```  should only show `libopencv_*_3.2` and not `libopencv_*_2.4`

**Comment by hitimo** [comment]
@yjmenezes  @RicardoRagel 
It should look like:
```
/media/timo/scnd/catkin_ws_aerial_mapper/devel/lib$ ldd libaerial_mapper_demos.so | grep opencv
        libopencv_core.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_core.so.3.2 (0x00007f2adc84c000)
        libopencv_imgproc.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_imgproc.so.3.2 (0x00007f2ada0f3000)
        libopencv_highgui.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_highgui.so.3.2 (0x00007f2ad9eb0000)
        libopencv_calib3d.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_calib3d.so.3.2 (0x00007f2ad9b1d000)
        libopencv_imgcodecs.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_imgcodecs.so.3.2 (0x00007f2ad892e000)
        libopencv_features2d.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_features2d.so.3.2 (0x00007f2ad53f9000)
        libopencv_xfeatures2d.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_xfeatures2d.so.3.2 (0x00007f2ad4f55000)
        libopencv_flann.so.3.2 => /media/timo/scnd/catkin_ws_aerial_mapper/devel/lib/libopencv_flann.so.3.2 (0x00007f2ad30c9000)
```

**Comment by RicardoRagel** [comment]
Hi @hitimo 

Yes, I have check that the executable pulls only opencv 3.2 (output looks like your output).
However,  `catkin build aerial_mapper` have some warnings:

`_______________________________________________________________________________________________________________________________________________________________________
Warnings   << aerial_mapper_demos:make /home/deltanet/catkin_ws_aerial_mapper/logs/aerial_mapper_demos/build.make.012.log                                              
/usr/bin/ld: warning: libopencv_stitching.so.3.2, needed by /home/deltanet/catkin_ws_aerial_mapper/devel/lib/libaerial_mapper_ortho.so, may conflict with libopencv_stitching.so.2.4
/usr/bin/ld: warning: libopencv_core.so.2.4, needed by /usr/lib/x86_64-linux-gnu/libopencv_stitching.so.2.4.8, may conflict with libopencv_core.so.3.2
/usr/bin/ld: warning: libopencv_stitching.so.3.2, needed by /home/deltanet/catkin_ws_aerial_mapper/devel/lib/libaerial_mapper_ortho.so, may conflict with libopencv_stitching.so.2.4
/usr/bin/ld: warning: libopencv_core.so.2.4, needed by /usr/lib/x86_64-linux-gnu/libopencv_stitching.so.2.4.8, may conflict with libopencv_core.so.3.2
/usr/bin/ld: warning: libopencv_stitching.so.3.2, needed by /home/deltanet/catkin_ws_aerial_mapper/devel/lib/libaerial_mapper_ortho.so, may conflict with libopencv_stitching.so.2.4
/usr/bin/ld: warning: libopencv_core.so.2.4, needed by /usr/lib/x86_64-linux-gnu/libopencv_stitching.so.2.4.8, may conflict with libopencv_core.so.3.2
/usr/bin/ld: warning: libopencv_stitching.so.3.2, needed by /home/deltanet/catkin_ws_aerial_mapper/devel/lib/libaerial_mapper_ortho.so, may conflict with libopencv_stitching.so.2.4
/usr/bin/ld: warning: libopencv_core.so.2.4, needed by /usr/lib/x86_64-linux-gnu/libopencv_stitching.so.2.4.8, may conflict with libopencv_core.so.3.2
cd /home/deltanet/catkin_ws_aerial_mapper/build/aerial_mapper_demos; catkin build --get-env aerial_mapper_demos | catkin env -si  /usr/bin/make --jobserver-fds=6,7 -j; cd -
......................................................................................................................................................................
`

So I suppose that some wrong dependencies exits :( 

In addition, I found that the program fails at:  /src/aerial_mapper/aerial_mapper_ortho/src/ortho-forward-homography.cc, line 201:

`blender_->prepare(rect)`

when `prepareBlenderForNextImage()` is called from the class constructor

**Comment by yjmenezes** [comment]
@hitimo 

Thanks for your kind reply, I will test it later.

I had gotten  same warnings as @RicardoRagel about libopencv version,

warning: libopencv_core.so.2.4, needed by /usr/lib/x86_64-linux-gnu/libopencv_stitching.so.2.4.8, may conflict with libopencv_core.so.3.2

**Comment by hitimo** [comment]
If you test, please use current master (incl. #24 )

**Comment by RicardoRagel** [comment]
@hitimo 

Problem fixed for my using the last commit ([1b62f1c](https://github.com/ethz-asl/aerial_mapper/commit/1b62f1c754f5849efcf18b3e024e47c62c258afe)).  Any openCV errors and good results at /tmp/result.jpg :)

Thanks a lot!

**Comment by hitimo** [comment]
@RicardoRagel Thanks for testing! Closing issue for now, feel free to reopen if sth. remains unsolved.
